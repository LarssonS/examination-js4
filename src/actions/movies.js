import database from '../firebase/firebase';

//ADD_MOVIE

export const addMovie = (movie) => ({
    type: 'ADD_MOVIE',
    movie 
});

export const startAddMovie = (movieData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            title = '',
            description = '',
            director = '',
            genre = '',
            filmWrapper = '',
            raiting = 0,
            createdAt = 0
        } = movieData;
        const movie = { title, description, director, genre, filmWrapper, raiting, createdAt, uid };

        return database.ref(`users/movies`).push(movie).then(() => {
            dispatch(addMovie({ ...movie }));
          });
    };
};

export const removeMovie = ({ id } = {}) => ({
    type: 'REMOVE_MOVIE',
    id
});

export const startRemoveMovie = ({ id } = {}) => {
    return(dispatch) => {
        return database.ref(`users/movies/${id}`).remove().then(() => {
            dispatch(removeMovie({ id }));
        });
    }
};

export const editMovie = ({id, updates}) => ({
    type: 'EDIT_MOVIE',
    id,
    updates
});

export const startEditMovie = ({id, updates}) => {
    return(dispatch) => {
        return database.ref(`users/movies/${id}`).update(updates).then(() => {
           dispatch(editMovie({id, updates})); 
        });
    };
};

export function fetchListenerMovies(ref, callback) {
    ref.on("child_added", (snapshot) => {
            const data = { ...snapshot.val(), id: snapshot.key };
            callback(data);
        });
};

export function removeListenerMovie(ref, callback) {
    ref.on("child_removed", (snapshot) => {
            callback(snapshot.key);
        });
};

export function changedListenerMovie(ref, callback) {
    ref.on("child_changed", (snapshot) => {
            const data = { updates: snapshot.val(), id: snapshot.key };
            callback(data);
        });
};
