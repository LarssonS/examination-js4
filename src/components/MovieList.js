import React from 'react';
import { connect } from 'react-redux';
import MovieListItem from './MovieListItem';
import selectMovies from '../selectors/movies';
import { Row, Col } from 'reactstrap';

const MovieList = (props) => (
    <div>
        <h1>Movie list</h1>
        
        <Row>
            {props.movies.length === 0 && <p>No movies</p>}
                {props.movies.map((movie) => {
                    return <Col sm="3" key={movie.id}> <MovieListItem {...movie} /> </Col> 
                })}
        </Row>
    </div>
);

const mapStateToProps = (state) => {
    return {
        movies: selectMovies(state.movies, state.filters)
    };
};

export default connect(mapStateToProps)(MovieList);