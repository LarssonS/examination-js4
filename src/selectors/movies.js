import moment from 'moment';

//Get visible movies
export default (movies, { text, sortBy, startDate, endDate }) => {
    return movies.filter((movie) => {
        const createdAtMoment = moment(movie.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        const titleMatch = movie.title.toLowerCase().includes(text.toLowerCase());
        const genreMatch = movie.genre.toLowerCase().includes(sortBy.toLowerCase());

        return startDateMatch && endDateMatch && titleMatch && genreMatch;
    });
}