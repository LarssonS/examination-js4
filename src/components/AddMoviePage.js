import React from 'react';
import { connect } from 'react-redux';
import MovieForm from './MovieForm';
import { startAddMovie } from '../actions/movies';
import { Container } from 'reactstrap';

export class AddMoviePage extends React.Component {
    onSubmit = (movie) => {
        this.props.startAddMovie(movie);
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
                <Container>
                    <h1>Add Movie</h1>
                    <MovieForm
                        onSubmit={this.onSubmit}
                    />
                </Container>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    startAddMovie: (movie) => dispatch(startAddMovie(movie))
});

export default connect(undefined, mapDispatchToProps)(AddMoviePage);