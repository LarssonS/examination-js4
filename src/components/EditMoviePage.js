import React from 'react';
import { connect } from 'react-redux';
import MovieForm from './MovieForm';
import { startEditMovie, startRemoveMovie } from '../actions/movies';
import { Container, Button } from 'reactstrap';

export class EditMoviePage extends React.Component {
    onSubmit = (movie) => {
        this.props.startEditMovie(this.props.movie.id, movie),
        this.props.history.push('/');
    };
    onRemove = () => {
        this.props.startRemoveMovie({ id: this.props.movie.id }),
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
                <Container>
                    <MovieForm 
                        movie={this.props.movie}
                        onSubmit={this.onSubmit}
                    />
                    <Button onClick={this.onRemove}>Remove</Button>
                </Container>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => {
    return {
        movie: state.movies.find((movie) => movie.id === props.match.params.id)
    };
};

const mapDispatchToProps = (dispatch) => ({
    startEditMovie: (id, movie) => dispatch(startEditMovie(id, movie)),
    startRemoveMovie: (data) => dispatch(startRemoveMovie(data))
});


export default connect(mapStateToProps, mapDispatchToProps)(EditMoviePage)