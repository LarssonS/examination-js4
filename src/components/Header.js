import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';
import {
  Navbar,
  Nav,
  NavItem,
  Button,
  Container
} from 'reactstrap';

export const Header = ({ startLogout }) => (
    <Navbar color="light" light expand="md">
        <Container>
            <Link to="/dashboard" className="text-dark h1">
                MovieHub
            </Link>
                <Nav className="ml-auto" navbar>
                    <NavItem className="mr-5">
                        <Link to="/create" className="text-dark h2">
                            Add a movie
                        </Link>
                    </NavItem>
                    <NavItem>
                        <Button className="btn-lg" onClick={startLogout}>Logout</Button>
                    </NavItem>
                </Nav>
        </Container>
    </Navbar>
);

const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);
