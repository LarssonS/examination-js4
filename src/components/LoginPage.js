import React from 'react';
import { connect } from 'react-redux';
import { startLogin, startOnLogin, startOnRegister } from '../actions/auth';
import { Button, Form, FormGroup, Label, Input, Jumbotron, Container,
        Navbar, NavbarToggler, NavbarBrand, Nav, NavItem, NavLink, Collapse } from 'reactstrap';

export class LoginPage extends React.Component {
    constructor(props) {
        super(props);
        
        this.showRegisterForm = this.showRegisterForm.bind(this);
        this.showLoginForm = this.showLoginForm.bind(this);

        this.state = {
            register: false,
            login: false,
            isOpen: false
        };
    }
    toggle = () => {
        this.setState(() => ({
            isOpen: !this.state.isOpen
        }));
    }
    startLogin = () => {
        this.props.startLogin();
    }
    showRegisterForm = () => {
        if (this.state.register === false) {
            this.setState(() => ({
                register: true,
                login: false
            }));
        } else {
            this.setState(() => ({
                register: false
            }));
        }
    }
    showLoginForm = () => {
        if (this.state.login === false) {
            this.setState(() => ({
                register: false,
                login: true
            }));
        } else {
            this.setState(() => ({
                login: false
            }));
        }
    }
    onRegister = (e) => {
        e.preventDefault();
        const email = e.target.email.value;
        const password = e.target.password.value;

        startOnRegister(email, password);
    }
    onLogin = (e) => {
        e.preventDefault();
        const email = e.target.email.value;
        const password = e.target.password.value;

        startOnLogin(email, password);
    }
    render() {
        return (
            <div>
                <Navbar color="light" light expand="md">
                    <Container>
                        <NavbarBrand>
                            <h1>MovieHub</h1>
                        </NavbarBrand>
                        <NavbarToggler onClick={this.toggle} />
                        <Collapse isOpen={this.state.isOpen} navbar>
                            <Nav className="ml-auto" navbar>
                                <NavItem className="ml-2">
                                    <Button className="btn-lg" onClick={this.showRegisterForm}>Register</Button>
                                </NavItem>
                                <NavItem className="ml-2">
                                    <Button className="btn-lg" onClick={this.showLoginForm}>Login</Button>
                                </NavItem>
                                <NavItem className="ml-2">
                                    <Button className="btn-lg" onClick={this.startLogin}>Login with Google</Button>
                                </NavItem>
                            </Nav>
                        </Collapse>
                    </Container>    
                </Navbar>

                {this.state.register && <Container className="mt-5"> <Form onSubmit={this.onRegister}>
                    <FormGroup>
                        <Label>Email</Label>
                        <Input type="text" name="email" className="form-control-lg" placeholder="something@email.com" />
                    </FormGroup>
                    <FormGroup>
                    <Label>Password</Label>
                        <Input type="password" name="password" className="form-control-lg" placeholder="Your secret" />
                    </FormGroup>      
                    <Button>Register</Button>
                </Form> </Container>}

                {this.state.login && <Container className="mt-5"> <form onSubmit={this.onLogin}>
                    <FormGroup>
                        <Label>Email</Label>
                        <Input type="text" name="email" className="form-control-lg" placeholder="something@email.com" />
                    </FormGroup>
                    <FormGroup>
                        <Label>Password</Label>
                        <Input type="password" name="password" className="form-control-lg" placeholder="Your secret" />
                    </FormGroup>
                    <Button>Login</Button>
                </form> </Container>}

                <Jumbotron className="h-100">
                    <Container className="w-50">
                        <h1 className="display-3">Welcome to MovieHub!</h1>
                        <h2>This is a simple movie blog!</h2>
                        <hr className="my-2" />
                        <p className="h2">Register and login. Or sign in with google!</p>
                    </Container>
                </Jumbotron>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginPage);
