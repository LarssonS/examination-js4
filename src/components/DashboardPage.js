import React from 'react';
import MovieList from './MovieList';
import MovieListFilters from './MovieListFilters';

const DashboardPage = () => (
    <div>
        <MovieListFilters />
        <MovieList />
    </div>
);

export default DashboardPage;
