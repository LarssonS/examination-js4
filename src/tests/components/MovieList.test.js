import React from 'react';
import { shallow } from 'enzyme';
import { MovieList } from '../../components/MovieList';
import movies from '../fixtures/movies';

test('should render MovieList with movies', () => {
  const wrapper = shallow(<MovieList movies={movies} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render MovieList with empty message', () => {
  const wrapper = shallow(<MovieList movies={[]} />);
  expect(wrapper).toMatchSnapshot();
});
