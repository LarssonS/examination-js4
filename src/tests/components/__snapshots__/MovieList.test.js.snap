// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render MovieList with empty message 1`] = `
<div>
  <h1>
    Movie list
  </h1>
  <Row
    tag="div"
  >
    <p>
      No movies
    </p>
  </Row>
</div>
`;

exports[`should render MovieList with movies 1`] = `
<div>
  <h1>
    Movie list
  </h1>
  <Row
    tag="div"
  >
    <Col
      className="mt-4"
      sm="3"
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
       
      <MovieListItem
        createdAt={0}
        description="superhero"
        director="Nolan"
        filmWrapper="img"
        genre="action"
        id="1"
        raiting={9}
        title="batman"
      />
       
    </Col>
    <Col
      className="mt-4"
      sm="3"
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
       
      <MovieListItem
        createdAt={-345600000}
        description="spider"
        director="secret"
        filmWrapper="img"
        genre="action"
        id="2"
        raiting={7}
        title="spiderman"
      />
       
    </Col>
    <Col
      className="mt-4"
      sm="3"
      tag="div"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    >
       
      <MovieListItem
        createdAt={345600000}
        description="superhero"
        director="unknown"
        filmWrapper="img"
        genre="sci-fi"
        id="3"
        raiting={6}
        title="superman"
      />
       
    </Col>
  </Row>
</div>
`;
